# HOMEWORK 1

*Sofiia Schipinskaia*

In [79]:
# importing the needed libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.feature_selection import SelectFromModel
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
import sklearn.metrics as metrics
from sklearn.model_selection import cross_validate, train_test_split
from sklearn.model_selection import GridSearchCV

## 1. Loading the data

In [8]:
# importing the file
df = pd.read_csv('dataset_simple.csv')
df.head(4)

feature_0  feature_1  feature_2  feature_3  feature_4  feature_5  \
0  -0.272674   0.801643  -0.650531   0.952486  -0.153037   0.427145   
1   0.201004  -0.263537  -0.793137  -4.483672   0.745123   1.156577   
2  -1.834495  -0.305209   0.053789   4.448289   1.294481   1.241900   
3  -0.484526   0.473056  -0.345463  -7.375136  -0.817676  -0.778446   

   feature_6  feature_7  feature_8  feature_9  ...    feature_991  \
0  -1.085534   1.318513   0.002588  -1.849832  ...      -0.572848   
1   0.351558   1.735511  -1.490988  -0.327063  ...       1.045496   
2  -1.358354  -0.211469  -0.186544   0.964812  ...       0.917945   
3   0.909734  -0.677147  -1.913900  -0.087426  ...       0.598949   

   feature_992  feature_993  feature_994  feature_995  feature_996  \
0    -0.448690     0.073369    -0.845503    -1.074903     0.222103   
1     1.494638    -0.776573    -1.219524    -0.756398    -0.963046   
2     0.920003     0.690022    -0.834381    -0.261806    -0.061880   
3     0.078195     0.024128     1.308711     0.943230     0.903181   

   feature_997  feature_998  feature_999  label  
0    -1.024658    -0.711647     0.036411      1  
1     0.488610     0.313609    -0.490794      0  
2    -1.177730    -1.345305    -0.085971      0  
3     0.501013     2.639802     1.540640      1  

[4 rows x 1001 columns]

In [4]:
# looking through the data
df.describe()

feature_0    feature_1    feature_2    feature_3    feature_4  \
count  9999.000000  9999.000000  9999.000000  9999.000000  9999.000000   
mean     -0.004460    -0.013556    -0.003732    -0.526323     0.005371   
std       0.991939     1.004086     1.004643     5.892651     0.995250   
min      -3.769974    -3.617617    -3.753391   -21.830355    -3.952088   
25%      -0.667193    -0.702984    -0.666734    -4.535838    -0.664072   
50%       0.006990    -0.019359     0.002527    -0.596745     0.008596   
75%       0.655618     0.667009     0.660033     3.367154     0.672424   
max       3.629974     4.437563     3.724971    20.021875     3.444207   

         feature_5    feature_6    feature_7    feature_8    feature_9  \
count  9999.000000  9999.000000  9999.000000  9999.000000  9999.000000   
mean     -0.008539     0.017156    -0.001695     0.016588    -0.011591   
std       1.001142     0.995826     1.000576     1.016175     0.994338   
min      -4.302776    -3.693491    -3.769295    -3.582291    -3.964497   
25%      -0.676919    -0.657419    -0.678497    -0.670545    -0.696125   
50%      -0.014751     0.023373    -0.005377     0.007411    -0.023052   
75%       0.662462     0.697545     0.662622     0.720524     0.668373   
max       3.502859     3.925565     4.069058     5.209648     3.871048   

          ...       feature_991  feature_992  feature_993  feature_994  \
count     ...       9999.000000  9999.000000  9999.000000  9999.000000   
mean      ...          0.004245     0.004588     0.000104    -0.004114   
std       ...          1.010292     0.996613     0.984661     1.005568   
min       ...         -3.543944    -3.702521    -3.645616    -3.757164   
25%       ...         -0.686178    -0.674913    -0.663270    -0.676386   
50%       ...         -0.002653     0.000651    -0.005459    -0.002117   
75%       ...          0.682821     0.680257     0.668761     0.675716   
max       ...          3.616481     4.456758     3.727838     3.959971   

       feature_995  feature_996  feature_997  feature_998  feature_999  \
count  9999.000000  9999.000000  9999.000000  9999.000000  9999.000000   
mean     -0.015524     0.003792    -0.012826    -0.000154     0.008867   
std       1.000405     0.992878     0.986676     1.003824     0.992667   
min      -3.998032    -3.844831    -4.314576    -3.912707    -3.852604   
25%      -0.690162    -0.660690    -0.671733    -0.689658    -0.653306   
50%      -0.017218     0.005911    -0.009118     0.024641    -0.002785   
75%       0.660026     0.680975     0.647070     0.674688     0.677321   
max       3.520052     4.305980     3.671758     3.555786     3.574397   

             label  
count  9999.000000  
mean      0.493149  
std       0.499978  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000  

[8 rows x 1001 columns]

In [6]:
# checking the sample imbalance problem 
df['label'].value_counts(normalize=True)

0    0.506851
1    0.493149
Name: label, dtype: float64

*As we can observe, the sample has no imbalance problem*

## 2. Feature selection

In [39]:
# splitting the data
X_train, X_test, y_train, y_test = train_test_split(df.drop('label', axis = 1), df['label'], test_size=0.2, random_state=45, shuffle=True)

In [40]:
# scaling the data
sc = StandardScaler()
sc.fit(X_train)
X_train_sc = sc.transform(X_train)
X_test_sc = sc.transform(X_test)

In [41]:
# cheking the shape of trainig sample
X_train_sc.shape

(7999, 1000)

In [42]:
# cheking the shape of test sample
X_test_sc.shape

(2000, 1000)

In [46]:
# feature selection
clf = ExtraTreesClassifier(n_estimators=100)
clf = clf.fit(X_train_sc, y_train) 
model = SelectFromModel(clf, prefit=True)
X_train_sc_fs = model.transform(X_train_sc)
X_test_sc_fs = model.transform(X_test_sc)

In [47]:
# cheking the shape of new trainig sample
X_train_sc_fs.shape

(7999, 163)

In [48]:
# cheking the shape of new test sample
X_test_sc_fs.shape

(2000, 163)

## 3. Building models and running cross validation

In [57]:
# identufying the function to estimate model performance
def model_performance(model_name, X_tr, y_tr, y_te, y_pred):
    print('Model name: %s'%model_name)
    print('Test accuracy (Accuracy Score): %f'%metrics.accuracy_score(y_te, y_pred))
    print('Test accuracy (ROC AUC Score): %f'%metrics.roc_auc_score(y_te, y_pred))
    print('Test precision 0: %f'%metrics.precision_score(y_te, y_pred, pos_label=0))   
    print('Test precision 1: %f'%metrics.precision_score(y_te, y_pred, pos_label=1)) 
    print('Test recall 0: %f'%metrics.recall_score(y_te, y_pred, pos_label=0))   
    print('Test recall 1: %f'%metrics.recall_score(y_te, y_pred, pos_label=1))

### Logistic Regression

In [88]:
# estimating the model on the training sample and checking performance 
clf_lr = LogisticRegression(random_state=0)
clf_lr.fit(X_train_sc_fs, y_train)
y_pred_lr = clf_lr.predict(X_test_sc_fs)
model_performance('Logistic Regression', X_tr = X_train_sc_fs, y_tr = y_train, y_te = y_test, y_pred = y_pred_lr)

Model name: Logistic Regression
Test accuracy (Accuracy Score): 0.824000
Test accuracy (ROC AUC Score): 0.823957
Test precision 0: 0.825616
Test precision 1: 0.822335
Test recall 0: 0.827246
Test recall 1: 0.820669


In [89]:
# cross-validation for Logistic Regression
scores_lr_f1 = cross_validate(clf_lr, X_train_sc_fs, y_train, cv=10, scoring=('f1'))
scores_lr_roc = cross_validate(clf_lr, X_train_sc_fs, y_train, cv=10, scoring=('roc_auc'))

In [90]:
# F1: results for LR
print('F1 scores: ', scores_lr_f1['test_score'])

F1 scores:  [0.86010363 0.82945736 0.82555283 0.8        0.85678705 0.81842105
 0.86848635 0.84169884 0.82097187 0.84130982]


In [91]:
# ROC AUC: results for LR
print('ROC AUC scores: ', scores_lr_roc['test_score'])

ROC AUC scores:  [0.93288645 0.90806884 0.90840556 0.88431128 0.92605836 0.89920411
 0.93331453 0.91878173 0.9046688  0.91452654]


### Decision Tree Classifier

In [74]:
# estimating the model on the training sample and checking performance 
clf_dt = DecisionTreeClassifier(random_state=0)
clf_dt.fit(X_train_sc_fs, y_train)
y_pred_dt = clf_dt.predict(X_test_sc_fs)
model_performance('Decision Tree Classifier', X_tr = X_train_sc_fs, y_tr = y_train, y_te = y_test, y_pred = y_pred_dt)

Model name: Decision Tree Classifier
Test accuracy (Accuracy Score): 0.655000
Test accuracy (ROC AUC Score): 0.654708
Test precision 0: 0.653956
Test precision 1: 0.656151
Test recall 0: 0.677196
Test recall 1: 0.632219


In [75]:
# cross-validation for Decision Tree
scores_dt_f1 = cross_validate(clf_dt, X_train_sc_fs, y_train, cv=10, scoring=('f1'))
scores_dt_roc = cross_validate(clf_dt, X_train_sc_fs, y_train, cv=10, scoring=('roc_auc'))

In [76]:
# F1: results for DT
print('F1 scores: ', scores_dt_f1['test_score'])

F1 scores:  [0.65087282 0.6835443  0.67421384 0.6898263  0.62548263 0.64419476
 0.63779528 0.65829146 0.64654088 0.6583229 ]


In [77]:
# ROC AUC: results for DT
print('ROC AUC scores: ', scores_dt_roc['test_score'])

ROC AUC scores:  [0.65057679 0.68783127 0.67667893 0.68810563 0.63596184 0.64346055
 0.65405465 0.6596478  0.64836435 0.6584477 ]


### Random Forest Classifier

In [78]:
# estimating the model on the training sample and checking performance 
clf_rf = RandomForestClassifier(n_estimators = 200, random_state = 0)
clf_rf.fit(X_train_sc_fs, y_train)
y_pred_rf = clf_rf.predict(X_test_sc_fs)
model_performance('Random Forest Classifier', X_tr = X_train_sc_fs, y_tr = y_train, y_te = y_test, y_pred = y_pred_rf)

Model name: Random Forest Classifier
Test accuracy (Accuracy Score): 0.893500
Test accuracy (ROC AUC Score): 0.893254
Test precision 0: 0.881679
Test precision 1: 0.906513
Test recall 0: 0.912142
Test recall 1: 0.874367


In [68]:
# cross-validation for Random Forest
scores_rf_f1 = cross_validate(clf_rf, X_train_sc_fs, y_train, cv=10, scoring=('f1'))
scores_rf_roc = cross_validate(clf_rf, X_train_sc_fs, y_train, cv=10, scoring=('roc_auc'))

In [69]:
# F1: results for RF
print('F1 scores: ', scores_rf_f1['test_score'])

F1 scores:  [0.9021164  0.91392405 0.91044776 0.86118252 0.90231362 0.88250653
 0.91370558 0.90566038 0.89203085 0.89698492]


In [70]:
# ROC AUC: results for RF
print('ROC AUC scores: ', scores_rf_roc['test_score'])

ROC AUC scores:  [0.96988527 0.97303735 0.96668953 0.94272931 0.96738016 0.95825343
 0.9714702  0.96114558 0.95308015 0.95770508]


## 4. Hyper parameters tuning

### Logistic Regression

In [92]:
# setting the parameters for logistic regression
grid_lr = {
    'penalty': ['l1', 'l2'], 
    'C': [0.001, 0.001, 0.01, 0.1, 1, 10, 100, 1000],
}

In [93]:
# applying grid search for LR
clf_grid_lr = GridSearchCV(clf_lr, grid_lr, scoring='roc_auc', cv=10)
clf_grid_lr.fit(X_train_sc_fs, y_train)

GridSearchCV(cv=10, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=0, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'penalty': ['l1', 'l2'], 'C': [0.001, 0.001, 0.01, 0.1, 1, 10, 100, 1000]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [94]:
# checking the best parameters and scores for LR
print(clf_grid_lr.best_estimator_)
print('Best parameters:', clf_grid_lr.best_params_)
print('Best ROC-AUC:', clf_grid_lr.best_score_)

LogisticRegression(C=0.1, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=0, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)
Best parameters: {'C': 0.1, 'penalty': 'l1'}
Best ROC-AUC: 0.9137553189791207


### Decision Tree Classifier

In [101]:
# setting the parameters for Decision Tree Classifier
grid_dt = {
    'max_depth': [100, 250, None], 
    'min_samples_leaf': [1, 2, 5]
}

In [102]:
# applying grid search for DT
clf_grid_dt = GridSearchCV(clf_dt, grid_dt, scoring='roc_auc', cv=10)
clf_grid_dt.fit(X_train_sc_fs, y_train)

GridSearchCV(cv=10, error_score='raise',
       estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=0,
            splitter='best'),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'max_depth': [100, 250, None], 'min_samples_leaf': [1, 2, 5]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [103]:
# checking the best parameters and scores for DT
print(clf_grid_dt.best_estimator_)
print('Best parameters:', clf_grid_dt.best_params_)
print('Best ROC-AUC:', clf_grid_dt.best_score_)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=100,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=5, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=0,
            splitter='best')
Best parameters: {'max_depth': 100, 'min_samples_leaf': 5}
Best ROC-AUC: 0.6936998676326697


### Random Forest Classifier

In [108]:
# setting the parameters for Random Forest Classifier
grid_rf = {
    'n_estimators': [100, 200, 300],
    'min_samples_leaf': [1, 2]    
}

In [109]:
# applying grid search for RF
clf_grid_rf = GridSearchCV(clf_rf, grid_rf, scoring='roc_auc', cv=10)
clf_grid_rf.fit(X_train_sc_fs, y_train)

GridSearchCV(cv=10, error_score='raise',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=200, n_jobs=1,
            oob_score=False, random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'n_estimators': [100, 200, 300], 'min_samples_leaf': [1, 2]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [110]:
# checking the best parameters and scores for RF
print(clf_grid_rf.best_estimator_)
print('Best parameters:', clf_grid_rf.best_params_)
print('Best ROC-AUC:', clf_grid_rf.best_score_)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=2, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=300, n_jobs=1,
            oob_score=False, random_state=0, verbose=0, warm_start=False)
Best parameters: {'min_samples_leaf': 2, 'n_estimators': 300}
Best ROC-AUC: 0.9663046744797972


## Results

As we can see, three models were biuld: logistic regression, decision tree and random forest. The models with default parameters have the following ROC-AUC measures:
- Roc-AUC of logic regression - 82.4%
- Roc-AUC of decision tree - 65.5%
- **Roc-AUC of random forest - 89.3%**

So, the model of random forest classifier has the best ROC-AUC.

To increase the accuracy of the models, the grid search technique was applyed. After applying we have the following ROC-AUC measures:
- Roc-AUC of logic regression - 91.37%
- Roc-AUC of decision tree - 69.36%
- **Roc-AUC of random forest - 96.63%**

Thus, the random forest classifier performs best out of these three model. It gives 96.63% of ROC-AUC with the parameters **{'min_samples_leaf': 2, 'n_estimators': 300}**.


*P.S. I used only this banch of parameters for the grid search part of the task because my laptop cannot compute the algorithms with larger set of parameters :(* 